### PR TITLE
Allow invalid field data to be kept along with the error messages in the Entity

### LINKED ISSUE #7615: 3.1: Don't forget about invalidated data in entities
I basically did
```php
$entities = $this->Examples->newEntities($array);

foreach ($entities as $entity) {
	$entityErrors = $entity->errors();
	if (!$entityErrors) {
		continue;
	}
	// Write error message of all fields incl input data that failed
```

But the entity does not contain the field data anymore (name and original_name):
```php
object(App\Model\Entity\Example) {
	'status' => (int) 2,
	'[new]' => true,
	'[accessible]' => [
		'*' => true
	],
	'[dirty]' => [
		'status' => true
	],
	'[original]' => [],
	'[virtual]' => [],
	'[errors]' => [
		'name' => [
			'isUnique' => 'valErrRecordNameExists'
		],
		'original_name' => [
			'isUnique' => 'valErrRecordNameExists'
		]
	]
}

```

I would have at least expected to have the failed field data available to make the error message clear where (what data) it failed, e.g.

```php
	'[invalid]' => [
		'name' => 'Foo',
		'original_name' => 'Bar',
	],
```

Which each [errors] entry of validation error messages it could then get the [invalid] entry of the data that was trying to be patched and make this the display error message in the view.

### PR DESCRIPTION
**Pull request by dereuromark**
Rebased  https://github.com/cakephp/cakephp/pull/7692 onto 3.2.
Mainly useful for mass validation/saving.

Resolves #7615

### CODE REVIEW

File: src/Datasource/RulesChecker.php [hunk 1]
```diff
@@ -327,6 +327,13 @@ protected function _addError($rule, $name, $options)
                 $message = [$message];
             }
             $entity->errors($options['errorField'], $message);
+
+            $invalid = null;
+            if (isset($entity->{$options['errorField']})) {
+                $invalid = $entity->{$options['errorField']};
+            }
+            $entity->invalid($options['errorField'], $invalid);
```
Comment (lorenzo) [inline src/Datasource/RulesChecker.php]: This line should be moved inside the if, where the check for the array property was done

Comment (dereuromark) [inline src/Datasource/RulesChecker.php]: You mean, we should not set a property here if null? Ok.

Comment (lorenzo) [inline src/Datasource/RulesChecker.php]: yes, sorry

**Comment by dereuromark** [comment]
All tests green :tada:

File: src/Datasource/InvalidPropertyInterface.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * @since         3.2.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Datasource;
+
+/**
+ * Describes the methods that any class representing a data storage should
+ * comply with.
+ */
+interface InvalidPropertyInterface
```
Comment (markstory) [inline src/Datasource/InvalidPropertyInterface.php]: :+1:

File: src/Datasource/RulesChecker.php [hunk 3]
```diff
@@ -327,6 +327,12 @@ protected function _addError($rule, $name, $options)
                 $message = [$message];
             }
             $entity->errors($options['errorField'], $message);
+
+            if (isset($entity->{$options['errorField']})) {
+                $invalidValue = $entity->{$options['errorField']};
+                $entity->invalid($options['errorField'], $invalidValue);
```
Comment (markstory) [inline src/Datasource/RulesChecker.php]: Shouldn't this have an interface check?

File: src/ORM/Marshaller.php [hunk 4]
```diff
@@ -128,6 +128,7 @@ public function one(array $data, array $options = [])
         $properties = [];
         foreach ($data as $key => $value) {
             if (!empty($errors[$key])) {
+                $entity->invalid($key, $value);
```
Comment (markstory) [inline src/ORM/Marshaller.php]: I think this and the other call below need interface checks.

File: src/Datasource/RulesChecker.php [hunk 5]
```diff
@@ -327,6 +327,12 @@ protected function _addError($rule, $name, $options)
...
+            }
+
             return $pass === true;
```
Comment (ADmad) [inline src/Datasource/RulesChecker.php]: Why not just `return $pass`?

**Comment by dereuromark** [comment]
Checks added.

File: src/ORM/Marshaller.php [hunk 6]
```diff
@@ -463,6 +467,9 @@ public function merge(EntityInterface $entity, array $data, array $options = [])
         $properties = $marshalledAssocs = [];
         foreach ($data as $key => $value) {
             if (!empty($errors[$key])) {
+                if (method_exists($entity, 'invalid')) {
```
Comment (markstory) [inline src/ORM/Marshaller.php]: interface check here too. :smiley_cat:

Comment (dereuromark) [inline src/ORM/Marshaller.php]: Derp ;)

**Comment by dereuromark** [comment]
:8ball:

**Comment by lorenzo** [comment]
Great job @dereuromark !

**Comment by dereuromark** [comment]
Thank you for bearing with me :) Biggest ORM contrubtion so far from my side :P

**Pull request merged**